File: 11/VMWriter.java
package comsys.compiler;

import comsys.compiler.enums.*;

import java.io.*;

public class VMWriter {

    BufferedWriter bufferedWriter;

    public VMWriter(String filename) {
        try {
            bufferedWriter = new BufferedWriter(new FileWriter(filename));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void writePush(Segment segment, int index) {
        writeFile("push " + segment.getValue() + " " + index);
    }

    public void writePop(Segment segment, int index) {
        writeFile("pop " + segment.getValue() + " " + index);
    }

    public void writeArithmetic(Command command) {
        writeFile(command.getValue());
    }

    public void writeLabel(String label) {
        writeFile("label " + label);
    }

    public void writeGoto(String label) {
        writeFile("goto " + label);
    }

    public void writeIf(String label) {
        writeFile("if-goto " + label);
    }

    public void writeCall(String name, int nArgs) {
        writeFile("call " + name + " " + nArgs);
    }

    public void writeFunction(String name, int nLocals) {
        writeFile("function " + name + " " + nLocals);
    }

    public void writeReturn() {
        writeFile("return");
    }

    void writeFile(String data) {
        try {
            bufferedWriter.write(data + "\n");
            bufferedWriter.flush();
        } catch(Exception e) {
            e.printStackTrace();
        }
    }

    public void close() {
        try {
            bufferedWriter.flush();
            bufferedWriter.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: 11_project/VMWriter.cpp
#include "VMWriter.h"

VMWriter::VMWriter(string vmFile) {
	writeFile.open(vmFile);
	tab = "";
}

void VMWriter::writePush(Segment segment, int index) {
	writeFile << tab << "push " << Segment_ele[segment] << " " << index << endl;
}

void VMWriter::writePop(Segment segment, int index) {
	writeFile << tab << "pop " << Segment_ele[segment] << " " << index << endl;
}

void VMWriter::writeArithmetic(Command command) {
	writeFile << tab << Command_ele[command] << endl;
}

void VMWriter::writeLabel(string label) {
	writeFile << tab << "label " << label << endl;
	tab += "  ";
}

void VMWriter::writeGoto(string label) {
	writeFile << tab << "goto " << label << endl;
}

void VMWriter::writeIf(string label) {
	writeFile << tab << "if-goto " << label << endl;
}

void VMWriter::writeCall(string name, int nArgs) {
	writeFile << tab << "call " << name << " " << nArgs << endl;
}

void VMWriter::writeFunction(string name, int nLocals) {
	writeFile << tab << "function " << name << " " << nLocals << endl;
	tab += "  ";
}

void VMWriter::writeReturn(void) {
	writeFile << tab << "return" << endl;
	tab.substr(0, tab.length() - 2);
}

void VMWriter::close(void) {
	writeFile.close();
}
